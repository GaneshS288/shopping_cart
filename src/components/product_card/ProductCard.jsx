import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";
import styles from "./ProductCard.module.css";
import starIcon from "../../assets/rating-star.svg";
import { useState } from "react";

function ProductCard({
  productData,
  handleAddToCart,
  handleRemoveFromCart,
  cart,
}) {
  const [quantity, setQuantity] = useState(1);
  const productCountInCart = getProductsInCartCount();

  function getProductsInCartCount() {
    const productInCart = cart.find((data) => data.id === productData.id);
    return productInCart ? productInCart.quantity : 0;
  }

  return (
    <div className={styles["productCard-container"]}>
      <div>
        <img
          src={productData.image}
          alt={productData.title}
          className={styles["product-img"]}
          loading="lazy"
        />
      </div>

      <h4 className={styles["product-title"]}>{productData.title}</h4>

      <div className={styles["priceAndRating-container"]}>
        <p>${productData.price}</p>
        <p>
          {productData.rating.rate}{" "}
          <img src={starIcon} alt="star icon" className={styles["star-icon"]} />
          ({productData.rating.count} reviews)
        </p>
      </div>

      <div className={styles["addToCart-container"]}>
        <button
          onClick={() =>
            handleAddToCart({ ...productData, quantity: quantity })
          }
          className={styles["addToCart-button"]}
        >
          Add to Cart
        </button>
        <div className={styles["quantity-container"]}>
          <button aria-label="decrement" className={styles["quantity-button"]}>
            -
          </button>
          <input
            type="text"
            value={quantity}
            onChange={(e) => setQuantity(Number(e.target.value))}
            name="quantity"
            aria-label="quantity"
            className={styles["quantity-input"]}
          />
          <button aria-label="increment" className={styles["quantity-button"]}>
            +
          </button>
        </div>
      </div>

      <div className={styles["detailsAndRemoveButton-container"]}>
        <button className={styles["details-button"]}>
          <NavLink
            to={`/home/products/${productData.id}`}
            className={styles["details-link"]}
          >
            View Details
          </NavLink>
        </button>

        <button
          type="button"
          onClick={() => handleRemoveFromCart(productData)}
          className={styles["remove-button"]}
        >
          Remove
        </button>
      </div>

      {productCountInCart > 0 ? (
        <p className={styles["product-count-in-cart"]}>({productCountInCart} in cart)</p>
      ) : null}
    </div>
  );
}

ProductCard.propTypes = {
  productData: PropTypes.object.isRequired,
  handleAddToCart: PropTypes.func.isRequired,
  handleRemoveFromCart: PropTypes.func.isRequired,
  cart: PropTypes.array.isRequired,
};

export { ProductCard };
