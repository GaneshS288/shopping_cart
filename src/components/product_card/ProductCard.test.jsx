import { describe, it, expect } from "vitest";
import { render, screen, act } from "@testing-library/react";
import { ProductCard } from "./ProductCard.jsx";
import { createMemoryRouter, RouterProvider } from "react-router-dom";


const dummyProductData = {
  category: "men's clothing",
  description:
    "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
  id: 1,
  image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
  price: 109.95,
  rating: { rate: 3.9, count: 120 },
  title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
};

function testSetup() {
  const routes = [
    {
      path : "/home/products",
      element: <ProductCard productData={dummyProductData}></ProductCard>,
    },
  ];

  const router = createMemoryRouter(routes, {
    initialEntries: ["/", "/home/products"],
    initialIndex: 1,
  });

  return { router }
}

describe("Product Card", () => {
  it("matches snapshot", () => {
    const { router } = testSetup();
    const {container} = render(<RouterProvider router={router}></RouterProvider>);
    
    expect(container).toMatchSnapshot();
  })

  it("renders the card", () => {
    const { router } = testSetup();
    render(<RouterProvider router={router}></RouterProvider>);
    screen.debug()
   

    const productImg = screen.getByRole("img");
    const productTitle = screen.getByRole("heading");
    const [addToCartButton, viewDetailsButton] = screen.getAllByRole("button");

    expect(productImg).toBeInTheDocument();
    expect(productTitle.textContent).toBe(dummyProductData.title);
    expect(addToCartButton.textContent).toBe("Add to Cart");
    expect(viewDetailsButton.textContent).toBe("View Details");
  });
});
