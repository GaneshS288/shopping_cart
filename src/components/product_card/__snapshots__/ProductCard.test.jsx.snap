// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Card > matches snapshot 1`] = `
<div>
  <div>
    <div>
      <img
        alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
        src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
      />
    </div>
    <h4>
      Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
    </h4>
    <button>
      Add to Cart
    </button>
    <button>
      <a
        class=""
        data-discover="true"
        href="/home/products/1"
      >
        View Details
      </a>
    </button>
  </div>
</div>
`;
